### PR TITLE
o/snapstate: allow updating snaps with components from local files

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This enables us to update snaps with components from local snap and component files.

### CODE REVIEW

File: overlord/snapstate/snapstate_update_test.go [hunk 1]
```diff
@@ -14446,3 +14436,374 @@ func (s *snapmgrTestSuite) TestRefreshCandidates(c *C) {
...
+		})
+	}
+
+	c.Assert(s.fakeBackend.ops.Ops(), DeepEquals, expected.Ops())
+	c.Assert(s.fakeBackend.ops, DeepEquals, expected)
+
+	// check progress
+	task := ts.Tasks()[1]
+
+	// verify snapSetup info
```
Comment (pedronis) [inline overlord/snapstate/snapstate_update_test.go:14741]: do we need to do similar checks for the component-setup? maybe this is a valid question also for other preexisting tests?

File: overlord/snapstate/target.go [hunk 2]
```diff
@@ -978,6 +978,14 @@ func UpdateWithGoal(ctx context.Context, st *state.State, goal UpdateGoal, filte
 		return nil, nil, err
 	}
 
+	for _, t := range plan.targets {
+		// sort the components by name to ensure we always install components in the
+		// same order
```
Comment (pedronis) [inline overlord/snapstate/target.go:983]: do we do the same for install?

File: overlord/snapstate/target.go [hunk 3]
```diff
@@ -1253,7 +1257,7 @@ func targetForPathSnap(update PathSnap, snapst SnapState, opts Options) (target,
 		return target{}, fmt.Errorf("cannot install local snap %q: %v != %v (revision mismatch)", update.InstanceName, update.RevOpts.Revision, si.Revision)
 	}
 
-	if update.RevOpts.Channel != "" && update.SideInfo.Channel != "" && update.RevOpts.Channel != update.SideInfo.Channel {
+	if update.RevOpts.Channel != "" && si.Channel != "" && update.RevOpts.Channel != si.Channel {
```
Comment (pedronis) [inline overlord/snapstate/target.go:1260]: what is this change update? is there a matching testt?

**Review by pedronis — COMMENTED** [review]
did a pass, some comments and some questions

File: overlord/snapstate/target.go (hunk 3 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:1260]: I think this is from me rebasing things, but `si == update.SideInfo`, it is declared up above.

File: overlord/snapstate/target_test.go [hunk 4]
```diff
@@ -419,7 +468,47 @@ components:
 	c.Check(snapsup.Channel, Equals, "edge")
 }
 
-func (s *TargetTestSuite) TestInstallFromPathRevOptsChannel(c *C) {
+func (s *snapmgrTestSuite) TestInstallComponentsFromPathInvalidComponentName(c *C) {
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target_test.go:471]: My mistake, fixed now.

File: overlord/snapstate/snapstate_update_test.go (hunk 1 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapstate_update_test.go:14763]: I added some checks in 93ed4b85b4de7453a6e6373f41f9abfa2c5a1936.j

File: overlord/snapstate/snapstate_install_test.go [hunk 5]
```diff
@@ -6977,87 +7001,3 @@ func printTasks(ts []*state.Task) string {
 	}
 	return b.String()
 }
-
-func (s *snapmgrTestSuite) TestInstallComponentsFromPathInvalidComponentFile(c *C) {
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/snapstate_install_test.go:6981]: Why are these two tests not needed anymore?

**Review by alfonsosanchezbeato — COMMENTED** [review]
I am a bit confused by the title/description of the PR. Is it to "refresh snaps" as in the title or to refresh components as in the description? Also, when talking about refreshing components from local files, is that for the case when we have assertions for those files? Otherwise it would not be a refresh, right? If that is the case maybe put that in the PR description.

**Comment by andrewphelpsj** [comment]
I've rebased this to hopefully make it a bit easier to review, since a lot has changed in master since this was originally written.

**Review by alfonsosanchezbeato — COMMENTED** [review]
The question I have about this is what this concretely enables. Is it being able to refresh snaps that have unasserted components? What would happen to the unasserted components in this case? We should have the discussion with @pedronis  of whether mixing asserted snaps with unasserted components or the inverse should be allowed at all.

**Comment by andrewphelpsj** [comment]
@alfonsosanchezbeato this essentially adds support for components to `UpdatePathWithDeviceContext`.

**Comment by andrewphelpsj** [comment]
Actually, this also adds a replacement for what `snapstate.InstallPathMany` provides right now, but for components. Note that `snapstate.InstallPathMany` is actually implemented as an update, so the naming can get confusing.

File: overlord/snapstate/snapstate_update_test.go [hunk 6]
```diff
@@ -14662,3 +14671,427 @@ func (s *snapmgrTestSuite) TestUpdateTasksWithComponentsRemoved(c *C) {
...
+	} else {
+		// make sure everything is back to how it started
+		c.Assert(snapst.Active, Equals, true)
+		c.Assert(snapst.Sequence.Revisions, HasLen, 1)
+		c.Assert(snapst.Sequence.Revisions[0], DeepEquals, seq.Revisions[0])
+	}
+}
+
+func checkComponentSetupTasks(c *C, ts *state.TaskSet, expected []snapstate.ComponentSetup) {
+	found := make([]snapstate.ComponentSetup, 0, len(expected))
```
Comment (pedronis) [inline overlord/snapstate/snapstate_update_test.go:15082]: shouldn't perhaps found and expected be a map between task kind and ComponentSetup ? are we seeing the same kind of multiple times here?

**Review by pedronis — COMMENTED** [review]
thank you, if I understand correctly this is mostly enabling and adding test for functionality that is already almost fully there on master? a test question though

File: overlord/snapstate/snapstate_update_test.go (hunk 6 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapstate_update_test.go:15082]: Yes, we're seeing the same kind multiple times. In this test, updating from a path, we see `prepare-component` tasks. In the regular refresh, we'd see `download-component` tasks.

I'll add a check that verifies which task they're found on.

**Comment by andrewphelpsj** [comment]
@pedronis, yes you're correct in stating that this mostly is just enabling code that is already in master and testing it. Much of the code is shared between the refresh and install cases.

**Review by pedronis — APPROVED** [review]
thank you

**Comment by andrewphelpsj** [comment]
Only failing test here is `google:ubuntu-20.04-64:tests/main/nvidia-files:535_server`, which is known.

**Pull request merged**